### PR TITLE
Egl yuv extension

### PR DESCRIPTION
**Pull request by erikhagglund**
review @emilwestergren

### CODE REVIEW

File: source/draw/gpu/opengl/AndroidContext.ooc [hunk 1]
```diff
@@ -165,6 +165,11 @@ AndroidContext: class extends OpenGLContext {
 		}
 		result
 	}
-	recycleEGLRgba: func (image: EGLRgba) { this _eglBin add(image) }
```
Comment (emilwestergren) [inline source/draw/gpu/opengl/AndroidContext.ooc]: This change is already up in a PR: https://github.com/magic-lang/ooc-kean/pull/1895

File: source/draw/gpu/opengl/EGLYuv.ooc [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+use geometry
+use draw
+use base
+use concurrent
+import GraphicBuffer, OpenGLContext, OpenGLPacked
+import backend/[GLTexture, GLContext, EGLImage]
+
+version(!gpuOff) {
+EGLYuv: class extends OpenGLPacked {
+	_buffer: GraphicBuffer
```
Comment (emilwestergren) [inline source/draw/gpu/opengl/EGLYuv.ooc]: Replace with the handle only

File: source/draw/gpu/opengl/EGLYuv.ooc [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import backend/[GLTexture, GLContext, EGLImage]
+
+version(!gpuOff) {
+EGLYuv: class extends OpenGLPacked {
+	_nativeBuffer: Pointer
+	nativeBuffer ::= this _nativeBuffer
+	init: func (_buffer: GraphicBuffer, context: OpenGLContext) {
+		super(EGLImage create(TextureType External, _buffer size, _buffer nativeBuffer, context backend), 3, context)
+	}
+	free: override func {
```
Comment (emilwestergren) [inline source/draw/gpu/opengl/EGLYuv.ooc]: You can remove this.

File: source/draw/gpu/opengl/shaders/unpackYuvToUv.frag [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#version 300 es
+#extension GL_EXT_YUV_target : enable
+precision mediump float;
+uniform __samplerExternal2DY2YEXT texture0;
+in highp vec2 fragmentTextureCoordinate;
+out vec2 outColor;
+void main() {
+	float uSample = texture(texture0, fragmentTextureCoordinate).g;
```
Comment (emilwestergren) [inline source/draw/gpu/opengl/shaders/unpackYuvToUv.frag]: Maybe a comment here why you sample twice

File: source/draw/gpu/opengl/YuvExtensionContext.ooc [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+		super(other)
+		this _yuvShader = OpenGLMapTransform new(slurp("shaders/yuv.frag"), this)
+		this _unpackY = OpenGLMap new(slurp("shaders/unpackYuv.vert"), slurp("shaders/unpackYuvToMonochrome.frag"), this)
+		this _unpackUv = OpenGLMap new(slurp("shaders/unpackYuv.vert"), slurp("shaders/unpackYuvToUv.frag"), this)
+		this _pack = OpenGLMap new(slurp("shaders/unpackYuv.vert"), slurp("shaders/packCompositeYuvToYuv.frag"), this)
+	}
+	free: override func {
+		this _backend makeCurrent()
+		this _eglBin free()
+		(this _yuvShader, this _unpackY, this _unpackUv, this _pack) free()
```
Comment (marcusnaslund) [inline source/draw/gpu/opengl/YuvExtensionContext.ooc]: You can merge this line and the one above

Comment (marcusnaslund) [inline source/draw/gpu/opengl/YuvExtensionContext.ooc]: Evaluation is from left to right, if it's important

File: source/draw/gpu/opengl/YuvExtensionContext.ooc [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+	preallocate: override func (resolution: IntVector2D) { this _eglBin clear() }
+	preregister: override func (image: Image) {
+		if (image instanceOf(GraphicBufferYuv420Semiplanar))
+			this createEGLYuv(image as GraphicBufferYuv420Semiplanar) free()
+	}
+	createEGLYuv: func (source: GraphicBufferYuv420Semiplanar) -> EGLYuv {
+		result := this _eglBin search(|eglYuv| source buffer nativeBuffer == eglYuv nativeBuffer)
+		if (result == null)
+			result = EGLYuv new(source buffer, this)
```
Comment (marcusnaslund) [inline source/draw/gpu/opengl/YuvExtensionContext.ooc]: `result := this _eglBin search(|eglYuv| source buffer nativeBuffer == eglYuv nativeBuffer) ?? EGLYuv new(source buffer, this)??`

**Comment by emilwestergren** [comment]
Rebase

File: source/draw/gpu/opengl/shaders/unpackYuvToUv.frag [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#version 300 es
+#extension GL_EXT_YUV_target : enable
+precision mediump float;
+uniform __samplerExternal2DY2YEXT texture0;
+in highp vec2 fragmentTextureCoordinate;
+out vec2 outColor;
+void main() {
+	// Sampling of u and v is done separetly as a workaround for a bug on an android device
```
Comment (emilwestergren) [inline source/draw/gpu/opengl/shaders/unpackYuvToUv.frag]: separately*

**Comment by erikhagglund** [comment]
no need :)

**Comment by erikhagglund** [comment]
Updated spelling

**Comment by marcusnaslund** [comment]
Looks good.

**Pull request merged**